### PR TITLE
NF: Gradient checking algorithm with fiber continuity

### PR DESCRIPTION
**Pull request by Atharva-Shah-2298**
## Description

This PR implements https://www.nature.com/articles/s41598-018-34940-4

## Motivation and Context

It solves the erroneous bval and bvec permutation issue. It uses flips and permutations and uses GFA from CSA 

## How Has This Been Tested?

Corrupted the hardii and sherbrooke bvecs and recovered successfully. All 17 tests cases related to the gradient file pass.

## Checklist

- [x ] I have read the [CONTRIBUTING](https://github.com/dipy/dipy/blob/master/.github/CONTRIBUTING.md) guidelines.
- [x ] My code follows the [DIPY coding style](https://docs.dipy.org/stable/devel/coding_style_guideline.html).
- [ x] I have added tests that cover my changes (if applicable).
- [x ] All new and existing tests pass locally.
- [ ] I have updated the documentation accordingly (if applicable).

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Maintenance / CI / Infrastructure

### CODE REVIEW

**Comment by maharshi-gor** [comment]
Hey @Atharva-Shah-2298 can you check this.

https://app.codecov.io/gh/dipy/dipy/pull/4004?dropdown=coverage&src=pr&el=h1

We have introduced new test cases but it does not cover everything that is introduced. Can you check the tests and update to cover this edge cases?

File: dipy/core/gradient_check.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+    if not np.any(mask):
+        raise ValueError(
+            "Fiber-continuity mask is empty. Provide an explicit `mask` or "
+            "loosen the default ADC/GFA thresholds."
+        )
+
+    csa = CsaOdfModel(gtab, sh_order_max=sh_order_max, smooth=smooth)
+    fit = csa.fit(data)
+    # Sample the ODF on the sphere for every voxel (we need the spatial
+    # gradient in voxels neighbouring the mask, so we sample everywhere).
```
Comment (maharshi-gor) [inline dipy/core/gradient_check.py]: This does not help creating further understanding.

**Comment by skoudoro** [comment]
Hi @Atharva-Shah-2298,

Thank you for this, I will test it later today or tomorrow. 

I see some refactoring needed (like `fit` called twice, why ? one time with mask, second time without mask, why ?), So I will try to look deeper into it and send a commit. 

Also, we need to add directly the CLI in this PR and why not a tutorial. 

Thanks again

**Comment by Atharva-Shah-2298** [comment]
Yeah! Refactoring needs to be done. Lot of raw code still. Testing out the methods on a corrupted data would be the first step! I will refactor lot of imports and placements and I will address codecov comments as well.

**Comment by maharshi-gor** [comment]
Thanks for this @Atharva-Shah-2298  I have tested with HARDI with corrputing bvecs from x,y,z -> y,-x,z

It worked for me.

**Review by skoudoro — CHANGES_REQUESTED** [review]
Hi @Atharva-Shah-2298,

Based on our discussion, did you have time to have a look on this, and fix the issue we saw? 

Thank in advance for the feedback

**Comment by Atharva-Shah-2298** [comment]
@skoudoro About why it corrects the already correct bvecs (No I haven been on to this). Regarding the comments from @maharshi-gor , I will address them when the method works as intended

From what I saw is that if the masking is wrong in the data, Background odfs are also making the algorithm off. So need a better mask for WM, I am thinking purely using GFA>0.3 because we need deep WM anyways

**Comment by skoudoro** [comment]
added the fiber coherence index algorithm from

```
Schilling et al. A fiber coherence index for quality control of B-table
    orientation in diffusion MRI scans. Magn Reson Imaging. 2019 May;58:82-89.
```

on this PR and will close #3447

I have also rebase the PR.

Now, the CLI will have  2 methods as an option

File: doc/examples/check_gradient_table.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+# ============================================================
+#
+# A single CLI command, ``dipy_correct_bvecs``, exposes both methods via the
+# ``--method`` flag.  The first positional argument is always the b-vector
+# file.
+#
+# **Aganj 2018** (default — needs only raw DWI)::
+#
+#     dipy_correct_bvecs path/to/dwi.bvec \
+#         --method aganj2018 \
```
Comment (Atharva-Shah-2298) [inline doc/examples/check_gradient_table.py:213]: method names will be clumsy if we add years of the published paper!

File: dipy/reconst/utils.py [hunk 3]
```diff
@@ -317,3 +319,154 @@ def convert_tensors(tensor, from_format, to_format):
...
+    """
+    permutations = list(itertools.permutations([0, 1, 2]))
+    nb_transforms = len(permutations) * nb_flips
+    transforms = np.zeros((nb_transforms, 3, 3), dtype=np.float32)
+
+    for i, perm in enumerate(permutations):
+        transforms[i * nb_flips][np.arange(3), perm] = 1.0
+        for axis in range(3):
+            flip_transform = transforms[i * nb_flips].copy()
+            flip_transform[axis, axis] *= -1
```
Comment (Atharva-Shah-2298) [inline dipy/reconst/utils.py:370]: This caused a bug! It flipped diagonal so only produced 1/2 of intended combinations. So just flip the entire row. Now fixed in the latest PR

**Review by Atharva-Shah-2298 — COMMENTED** [review]
Changed a few things for a small BF. My comment above about selective flips is still valid. Tested it thoroughly that subset of flips wont be corrected as overall ODF structure would remain intact

**Comment by maharshi-gor** [comment]
Hey @Atharva-Shah-2298 what is the status of this PR?

As we are planning to create a release end of sept / early oct.

Please let us know we should spend time on exploring this method.

**Comment by Atharva-Shah-2298** [comment]
> Hey @Atharva-Shah-2298 what is the status of this PR?
> 
> As we are planning to create a release end of sept / early oct.
> 
> Please let us know we should spend time on exploring this method.

I think I pushed all the changes. @skoudoro was working on it. Let me know if you need anything from me. Happy to helping in expediting the release

**Comment by maharshi-gor** [comment]
You might need to rebase the PR with master @Atharva-Shah-2298

**Comment by maharshi-gor** [comment]
I will take a re-look into it and we can decide after that.
